python/paddle/v2/framework/tests/op_test.py

```diff
@@ -5,175 +5,313 @@
+                jacobian_matrix.resize(
+                    (tensor_size, len(output_vals)), refcheck=False)
+                jacobian_matrix[row, :] = output_vals.flatten()
+            else:
+                start_idx = 0
+                for output_name in output_names:
+                    op.run(scope, ctx)
+                    output_val = np.array(
+                        cls.get_tensor_by_name(scope, output_name)).flatten()
+                    jacobian_matrix[row, start_idx:start_idx+len(output_val)] \
+                            = output_val
+                    start_idx += len(output_val)
+
+        for i in xrange(tensor_size):
+            if in_place:
+                cls.feed_input(scope, op, inputs, cpu_place)
+            origin_val = tensor_to_check.get_float_element(i)
+            x_pos = origin_val + delta
+            tensor_to_check.set_float_element(i, x_pos)
+            concat_flatten_output(i, x_pos_jacobian)
+            if in_place:
+                cls.feed_input(scope, op, inputs, cpu_place)
+            x_neg = origin_val - delta
+            tensor_to_check.set_float_element(i, x_neg)
+            concat_flatten_output(i, x_neg_jacobian)
+            tensor_to_check.set_float_element(i, origin_val)
+
+        grad_jacobian = (x_pos_jacobian - x_neg_jacobian) / delta / 2
+        # return numeric gradient jacobian matrix
+        if strict == False:
+            return grad_jacobian.sum(axis=1).reshape(tensor_to_check.shape())
+        return grad_jacobian
+
+    # TODO(pkuyym) should pass output_names not outputs
+    @classmethod
+    def get_simple_analytic_grads(cls,
+                                  scope,
+                                  op,
+                                  inputs,
+                                  outputs,
+                                  grad_name,
+                                  place,
+                                  no_grad_set=None):
+        ctx = core.DeviceContext.create(place)
+        cls.feed_input(scope, op, inputs, place)
+        # run forward
+        op.infer_shape(scope)
+        op.run(scope, ctx)
 
-    backward_op.infer_shape(scope)
-    backward_op.run(scope, ctx)
+        if no_grad_set is None:
+            no_grad_set = set()
 
-    out = np.array(scope.find_var(grad_name).get_tensor())
-    return out
+        backward_op = cls.get_backward_op(scope, op, no_grad_set)
+        # feed Input(Out@Grad), just set to one for all values
+        for out_var_name, is_dup in Operator.get_op_outputs(op.type()):
+            if out_var_name in outputs:
+                if is_dup:
+                    sub_vars = outputs[out_var_name]
+                    for sub_var_name, _ in sub_vars:
+                        out_var_tensor = cls.get_tensor_by_name(scope,
+                                                                sub_var_name)
+                        data = np.ones(out_var_tensor.shape(), dtype=np.float64)
```

Comment:
Is `np.float64` ok for op using float32?

Reply:
Since the unittest works well, I think passing np.float64 to operator with float32 is ok.